confirm.lua

```diff
@@ -0,0 +1,535 @@
+--@ module = true
+
+local gui = require('gui')
+local json = require('json')
+local overlay = require('plugins.overlay')
+local widgets = require("gui.widgets")
+
+local CONFIG_FILE = 'dfhack-config/confirm.json'
+
+registry = registry or {}
+
+------------------------
+-- Confirmation configs
+
+ConfirmConf = defclass(ConfirmConf)
+ConfirmConf.ATTRS{
+    id=DEFAULT_NIL,
+    title='DFHack confirm',
+    message='Are you sure?',
+    intercept_keys={},
+    context=DEFAULT_NIL,
+    predicate=DEFAULT_NIL,
+    pausable=false,
+    intercept_frame=DEFAULT_NIL,
+}
+
+function ConfirmConf:init()
+    if not self.id then
+        error('must set id to a unique string')
+    end
+    if type(self.intercept_keys) ~= 'table' then
+        self.intercept_keys = {self.intercept_keys}
+    end
+    for _, key in ipairs(self.intercept_keys) do
+        if key ~= '_MOUSE_L' and key ~= '_MOUSE_R' and not df.interface_key[key] then
+            error('Invalid key: ' .. tostring(key))
+        end
+    end
+    if not self.context then
+        error('context must be set to a bounding focus string')
+    end
+
+    if registry[self.id] then
+        error('id already registered: ' .. tostring(self.id))
+    end
+
+    -- auto-register
+    registry[self.id] = self
+end
+
+local function trade_goods_selected()
+    local function goods_selected(vec)
+        for _, sel in ipairs(vec) do
+            if sel == 1 then return true end
+        end
+    end
+
+    return goods_selected(df.global.game.main_interface.trade.goodflag[0]) or
+        goods_selected(df.global.game.main_interface.trade.goodflag[1])
+end
+
+local function has_caravans()
+    for _, caravan in pairs(df.global.plotinfo.caravans) do
+        if caravan.time_remaining > 0 then
+            return true
+        end
+    end
+end
+
+if not next(registry) then
```

Comment:
I think we should try to avoid checks like this because it makes development harder. Making any changes to the `ConfirmConf` invocations below would require manually clearing `registry` or restarting DF. Previously I think reloading the plugin (or the plugin lua module) was enough to apply changes.

Reply:
would it be better to take out the line that checks for duplicate IDs in `ConfirmConf:init()`? That was intended to catch copy paste errors, but if it were taken out, then the conf definitions would get automatically refreshed on every run

Reply:
suggested change in https://github.com/DFHack/scripts/pull/915/commits/1bf799631920940b017079e055d7ee6431158e2c